File: Hackerrank/Ad Infinitum - Math Programming Contest March'14/A Very Special Multiple.py
import fractions

maxN = 100000
isPrime = [True] * (maxN + 1)
isPrime[0] = isPrime[1] = False
prime = []
for i in range(2, maxN + 1):
    if isPrime[i]:
        for j in range(i + i, maxN + 1, i):
            isPrime[j] = False
        prime.append(i)

def check(x, l):
    rem, p, e = 0, 4, 10
    while l:
        if (l & 1) == 1:
            rem = (rem * e + p) % x
        p = (p * e + p) % x
        e = (e * e) % x
        l >>= 1
    return rem == 0

def solve(x):
    y = x
    c2, c5 = 0, 0
    while x % 2 == 0:
        c2 += 1
        x //= 2
    while x % 5 == 0:
        c5 += 1
        x //= 5
    zero = max(max(c2 - 2, 0), c5)
    p, c = [], []
    for i in range(len(prime)):
        if prime[i] * prime[i] > x or x <= 1:
            break
        if x % prime[i] == 0:
            cnt = 0
            p.append(prime[i])
            while x % prime[i] == 0:
                x //= prime[i]
                cnt += 1
            c.append(cnt)
    if x > 1:
        p.append(x)
        c.append(1)
    m = [0] * len(p)
    for i in range(0, len(p)):
        if p[i] == 3:
            m[i] = 3
            continue
        p1 = int(p[i] - 1)
        p2 = int(p1 ** 0.5)
        m[i] = p1
        flag = False
        for j in range(2, p2 + 1):
            if p1 % j == 0:
                flag = check(p[i], j)
                if flag:
                    m[i] = j
                    break
        if not flag:
            for j in range(p2 + 1, 1, -1):
                if p1 % j == 0:
                    flag = check(p[i], p1 // j)
                    if flag:
                        m[i] = p1 // j
                        break
    for i in range(len(p)):
        m[i] *= int(p[i]) ** int(c[i] - 1)
    four = 1
    if len(m) > 0:
        four = m[0]
    for i in range(1, len(p)):
        four = (four * m[i]) / fractions.gcd(four, m[i])
    return int(four) * 2 + zero

T = int(input())
for _ in range(T):
    x = int(input())
    print(solve(x))

File: Hackerrank/Ad Infinitum - Math Programming Contest March'14/A Very Special Multiple.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <queue>
#include <algorithm>

const int maxn = 100000;
bool isNotPrime[maxn + 1] = {true, true};
int F[20] = {0}, E[20] = {1};
int prime[maxn / 2], idx = 0;
void sieve() {
    for (int i = 2; i <= maxn; ++i) 
        if (!isNotPrime[i]) {
            for (int j = i + i; j <= maxn; j += i) isNotPrime[j] = true;
            prime[idx++] = i;
        }
}

void solve(long long x) {
    long long t = x;
    int c2 = 0, c5 = 0;
    while (t % 2 == 0) {
        t /= 2;
        c2 ++;
    }
    while (t % 5 == 0) {
        t /= 5;
        c5 ++;
    }
    long long zero = std::max(c5, std::max(c2 - 2, 0)), four = 1, rem = 4 % t;
    for (int i = 1; ; ++i) {
        if (rem == 0) {
            std::cout << "               " << x << " " << four << " " << zero << std::endl;
            if (std::sqrt(x) - four >= 0)
                std::cout << std::sqrt(x) << " " << four << std::endl;
            // std::cout << four * 2 + zero << std::endl;
            break;
        }
        rem = (rem * 10 + 4) % t;
        four ++;
        if (i % 100000000 == 0) std::cout << i << std::endl;
    }
}
bool check(long long x, long long len) {
    long long l = len, rem = 0;
    while (l) {
        if (l >= 19) {

        }
    }
    return rem == 0;
}
void solve2(long long x) {    
    long long t = x;
    int c2 = 0, c5 = 0;
    while (t % 2 == 0) {
        t /= 2;
        c2 ++;
    }
    while (t % 5 == 0) {
        t /= 5;
        c5 ++;
    }
    long long zero = std::max(c5, std::max(c2 - 2, 0)), four = 1, rem = 4 % t;
}
int T;
int main() {
    sieve();
    // freopen("in", "r", stdin);
    // std::cin >> T;
    // while (T--) {
    //     long long x;
    //     std::cin >> x;
    //     solve(x);
    // }
    for (int i = 0; i <= 100; ++i) {
        bool flag = true;
        if (flag) solve(prime[i]);
    }
    return 0;
}